setuptools/config.py

```diff
@@ -594,6 +595,22 @@ def parsers(self):
             'entry_points': self._parse_file,
             'py_modules': parse_list,
             'python_requires': SpecifierSet,
+            'cmdclass': parse_cmdclass,
+        }
+
+    def _parse_cmdclass(self, value):
+        def resolve_class(qualified_class_name):
+            idx = qualified_class_name.rfind('.')
+            class_name = qualified_class_name[idx+1:]
+            pkg_name = qualified_class_name[:idx]
+
+            module = __import__(pkg_name)
```

Comment:
Couple of things here. First, `__import__` as used here will always import the top-level module, so if `pkg_name` was `mypkg.mymod`, `pkg_name` would end up being `mypkg` (with `mymod` imported and a property of `mypkg`). Second `importlib.import_module` is preferred over the discouraged `__import__` _and_ it has the nice property of returning the last item, addressing the first concern, so I'd use that.

Reply:
Hi Jason,

according to https://setuptools.readthedocs.io/en/latest/references/keywords.html the keyword cmdclass is defined as: "A dictionary providing a mapping of command names to Command subclasses".

but such keyword is not listed in https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html

My interpretation is that declarative config docs are not complete yet. Perhaps I'm wrong, actually I discovered declarative config just a week ago and I'm upgrading my projects to adopt it.

Coming to the use case, my need is to generate some sources before the actual (standard) project build. Eg. embed static js generated files from a node application, thus launching npm build in a subfolder, or compile GRPC proto files as in https://stackoverflow.com/a/54236739.

Before the declarative config the common solution was to derive build_py command. Furthermore, the cmdclass keyword was there to provide custom commands and I thought that declarative config was on the way to support it as well.

I don't understand what you mean by "namely by having another package that presents a distutils.commands entry point".

Finally, I thought setuptools still supported python 2. Later I discovered that p2 support has been dropped with 53.0.0. So yes, resolve_class should be rewritten using importlib.

Reply:
Note: cmdclass is a dict from command name to Commad class, ie. fully qualified class names in setup.cfg.
Example, setup.cfg:

```
cmdclass =
  mycustomcmd  = mycustommodule.CustomCommand
```

where mycustommodule is a file python module in the folder containing setup.cfg too:

```
from setuptools import Command

class CustomCommand(Command):
    ...
```

Reply:
> My interpretation is that declarative config docs are not complete yet.

Another interpretation was that this feature was determined to be unnecessary for declarative config. I don't recall if it was an intentional omission, oversight, or deferred feature.

> Coming to the use case, my need is to generate some sources before the actual (standard) project build. Eg. embed static js generated files from a node application, thus launching npm build in a subfolder, or compile GRPC proto files as in https://stackoverflow.com/a/54236739.

Thanks for sharing the use-case. I can't help but feel like the approach Setuptools presents here is sub-optimal, mainly because the caller is required to override the `build_py` command just to inject a behavior before building.

I'm not sure Setuptools needs to continue to encourage this behavior and should perhaps provide a better interface for projects to inject behavior at various stages, e.g. to allow a project to declare "run 'grpc' before 'build'".

That said, I don't have an obvious solution for this approach, so perhaps it makes sense to honor this setting in setup.cfg.